flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,433 @@
+					if (element.isRecord()) {
+						StreamRecord<IN2> record = element.asRecord();
+						synchronized (lock) {
+							numRecordsIn.inc();
+							streamOperator.setKeyContextElement2(record);
+							streamOperator.processElement2(record);
+							inputSelection = inputSelector.nextSelection();
+						}
+
+						continue;
+					} else if (element.isWatermark()) {
+						Watermark watermark = element.asWatermark();
+						synchronized (lock) {
+							input2WatermarkGauge.setCurrentWatermark(watermark.getTimestamp());
+							streamOperator.processWatermark2(watermark);
+						}
+						continue;
+					} else {
+						auxiliaryHandler.processOtherElement2(element);
+						continue;
+					}
+				}
+			}
+
+			/* no data or the input is finished */
+			if (!setUnavailableAndListenInput(readingInputIndex)) {
+				// if the input is no longer listenable, then it is finished
+				if (finishInput(readingInputIndex)) {
+					// all inputs have been finished, exit
+					return false;
+				}
+			}
+		}
+
+		return true;
+	}
+
+	@VisibleForTesting
+	public final boolean isContinuousProcessing() {
+		return continuousProcessing;
+	}
+
+	@VisibleForTesting
+	public final void setContinuousProcessing(boolean continuousProcessing) {
+		this.continuousProcessing = continuousProcessing;
+	}
+
+	/**
+	 * It is called by another thread, so it must be thread-safe.
+	 */
+	public void stop() {
+		continuousProcessing = false;
+
+		// cancel all inputs and activate the main loop if it is blocked
+		for (Input input : new Input[]{input1, input2}) {
+			input.unlisten();
+		}
+	}
+
+	public void cleanup() {
+		// close all inputs
+		for (Input input : new Input[]{input1, input2}) {
+			try {
+				input.close();
+			} catch (Throwable t) {
+				LOG.warn("An exception occurred when closing input" + input.getId() + ".", t);
+			}
+		}
+	}
+
+	private boolean setUnavailableAndListenInput(int inputIndex) {
+		int bitMask = 1 << inputIndex;
+
+		// set `unavailable` flag
+		availableInputsMask.getAndUpdate(mask -> mask & (~bitMask));
+
+		// listen the input
+		Input input = inputIndex == 0 ? input1 : input2;
+		if (!input.isFinished()) {
+			listenFutures[inputIndex] = input.listen().thenApply(v -> {
+				availableInputsMask.getAndUpdate(mask -> mask | bitMask);
+				return inputIndex;
+			});
+
+			return true;
+		}
+
+		return false;
+	}
+
+	private void waitForAvailableInput(int selectionMask) throws IOException, ExecutionException, InterruptedException {
+		if (selectionMask == InputSelection.ALL.getInputMask() || selectionMask == TWO_INPUT_ANY_MASK) {
+			List<CompletableFuture<Integer>> futures = new ArrayList<>();
+			for (CompletableFuture<Integer> future : listenFutures) {
+				if (future != null) {
+					futures.add(future);
+				}
+			}
+			checkState(futures.size() > 0, "No input in listening.");
+
+			// block to wait for a available input
+			CompletableFuture.anyOf(futures.toArray(new CompletableFuture<?>[futures.size()])).get();
```

Comment:
As commented in other places, I suggest to pass an array of size 0 to `toArray`.